### PR TITLE
core: Add tests for many directives

### PR DESCRIPTION
**Pull request by jtszalay**
Contributes to https://github.com/mohammed90/caddy-ngrok-listener/issues/10

Adds testing for all parsing of options currently available.

Edit: Solved that test!
~There is a todo to fix an assert on a basicauth test:~
~https://github.com/jtszalay/caddy-ngrok-listener/blob/03-15-Test_logging_and_formatting/http_test.go#L71-L100~
~I'm currently chatting with Josh of the ngrok-go team (Hi, Josh!) about how best to enable that test to work.~

### CODE REVIEW

**Comment by jtszalay** [comment]
I'm going to focus on adding some more robust tests before the new directives and features.

**Comment by jtszalay** [comment]
Whew! There's a lot here but look at this: 

```bash
~ make coverage                  
go test -coverprofile cover.out ./...
ok      github.com/mohammed90/caddy-ngrok-listener      0.640s  coverage: 97.9% of statements
go tool cover -html cover.out
```

**Comment by jtszalay** [comment]
This and the others are ready for review. No rush on any of it :)

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ifeq (,$(shell go env GOBIN))
+GOBIN=$(shell go env GOPATH)/bin
+else
+GOBIN=$(shell go env GOBIN)
+endif
+
+# Targets
+
+.PHONY: all
+all: build
```
Comment (mohammed90) [inline Makefile:11]: I don't see the `build` target. Is `build` target left by mistake or is it an intrinsic target in make? I'm not aware of it being the latter 🤔

File: http.go [hunk 2]
```diff
@@ -43,11 +43,16 @@ type HTTP struct {
...
-	BasicAuth map[string]string `json:"basic_auth,omitempty"`
+	BasicAuth []basicAuthCred `json:"basic_auth,omitempty"`
 
 	l *zap.Logger
 }
 
+type basicAuthCred struct {
+	username string
+	password string
+}
```
Comment (mohammed90) [inline http.go:54]: JSON config cannot be marshalled into unexported fields

File: ngroklistener_test.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		err = tun.Provision(ctx)
+
+		if gt.expectProvisionErr {
+			require.NotNil(t, err)
+			return
+		} else {
+			require.Nil(t, err)
+			ngrokTun := tun.NgrokTunnel()
+			require.Equal(t, gt.expectedOpts, ngrokTun)
+		}
```
Comment (mohammed90) [inline ngroklistener_test.go:62]: You're already `return`ing in the `if`-branch, so you can continue without having to use `else`.

```suggestion
		}
		require.Nil(t, err)
		ngrokTun := tun.NgrokTunnel()
		require.Equal(t, gt.expectedOpts, ngrokTun)
```

**Review by mohammed90 — CHANGES_REQUESTED** [review]
This is so cool! I like what you've done with generics and the coverage percentage. Thank you for that! There are few minor comments.

File: Makefile (hunk 1 above)
Comment (jtszalay) [reply Makefile:11]: Whoops! Yep meant to remove that one! Thank you!

**Review by mohammed90 — APPROVED** [review]
Thank you!

**Pull request merged**